docs/release-notes/mongoid-8.1.txt

```diff
@@ -145,3 +145,45 @@ to or from a certain value:
   # => true
   person.name_changed?(from: "Trout", to: "Fletcher")
   # => false
+
+
+Added non-block form to ``#with`` and added ``#with!`` method
+-------------------------------------------------------------
+
+Mongoid 8.1 adds support for a non-block form for the ``#with`` method and for
+a new ``with!`` method. The ``#with`` method is used for changing the
+``persistence context`` of a Model class, a document, and, starting in Mongoid
+8.1, a criteria.
+
+Note that, ``with`` cannot be called on document instances without a block, since
+in order to change the persistence context of a document instance, the document
+itself must be modified. This is where the ``#with!`` method comes in: The
+``#with!`` method modifies its caller, and returns it with the given
+persistence context. This method can be used on document and criteria instances.
```

Comment:
My two cents: it doesn't make a lot of sense to have a block form of `.with` on a document instance. AFAIK, `band.with { ... }` is 100% equivalent to `band.class.with { ... }` (i.e. `Band.with { ... }`)

If my understanding above is correct, I would advocate to get rid of the instance block syntax, as it redundant/confusing, and instead allow only the non-block form on the instance `band.with()` -- which is what you currently have as the `.with!` method.

Reply:
I would be okay with deprecating document instance with, what do you think @p-mongo ?

Reply:
We decided that if instance-level #with didn't exst, we would not implement it now because it is in practice probably attainable via `band.class.with { ... }`, but since the functionality is already implemented, and it is not currently causing issues, let's leave it as is.